File: 295-find-median-from-data-stream/find-median-from-data-stream.java
class MedianFinder {
     Queue<Integer> minheap;
     Queue<Integer> maxheap;
    public MedianFinder() {
        minheap = new PriorityQueue<Integer>();
        maxheap = new PriorityQueue<Integer>(Collections.reverseOrder());

    }
    
    public void addNum(int num) {

        minheap.add(num);

        maxheap.add(minheap.peek());
        minheap.poll();

        if (maxheap.size() > minheap.size() + 1){
            minheap.add(maxheap.poll());
        }

      

    }
    
    public double findMedian() {
        
        if (minheap.size() == maxheap.size()){
            return (double)((minheap.peek() + maxheap.peek())*0.5);
        }
        else if (minheap.size() > maxheap.size()){
            return minheap.peek();
        }
        else return maxheap.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295-find-median-from-data-stream/find-median-from-data-stream.py
class MedianFinder:
    
    def __init__(self):
        self.minheap = []
        self.maxheap = []

    def addNum(self, num: int) -> None:

        # if not self.maxheap:
        #     heapq.heappush(self.maxheap, -num)
        #     return None
        
        heapq.heappush(self.maxheap, -num)
        heapq.heappush(self.minheap, -heapq.heappop(self.maxheap))

        if len(self.minheap) - len(self.maxheap) > 1:
            heapq.heappush(self.maxheap, -heapq.heappop(self.minheap))
        
    def findMedian(self) -> float:

        if len(self.minheap) == len(self.maxheap):
            return (-self.maxheap[0] + self.minheap[0]) / 2.0

        return self.minheap[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
